core/reorder/mc64.cpp

```diff
@@ -0,0 +1,615 @@
+            }
+
+            const ValueType dnew = lsp + weights[idx] - dual_u[col] - dual_vi;
+
+            if (dnew < lsap) {
+                if (ip[col] == -1) {
+                    // col is unmatched so we found an augmenting path.
+                    lsap = dnew;
+                    jsap = col;
+                    parents[col] = row;
+                } else {
+                    if ((gen != root || dnew < distance[col]) &&
+                        gen != -num_rows - root) {
+                        distance[col] = dnew;
+                        parents[col] = row;
+                        if (abs(dnew - lsp) < tolerance) {
+                            // dnew is the shortest currently possible distance,
+                            // so col can be put into q_j and be marked
+                            // accordingly.
+                            generation[col] = -num_rows - root;
+                            q_j.push_back(col);
+                        } else if (gen != root) {
+                            // col was not encountered before.
+                            generation[col] = root;
+                            Q.insert(dnew, col);
+                        } else {
+                            // col was already encountered but with larger
+                            // distance on a different path.
+                            generation[col] = root;
+                            Q.update_key(dnew, col);
+                        }
+                    }
+                }
+            }
+        }
+    }
+    if (lsap != inf) {
+        IndexType col = jsap;
+        // Update the matching along the shortest augmenting path.
+        do {
+            row = parents[col];
+            ip[col] = row;
+            auto idx = row_ptrs[row];
+            while (col_idxs[idx] != col) {
+                idx++;
+            }
+            idxs[row] = idx;
+            std::swap(col, p[row]);
+        } while (row != root);
+        // Update the dual vector u.
+        for (size_type i = 0; i < marked_counter; i++) {
+            const auto col = marked_cols[i];
+            dual_u[col] += distance[col] - lsap;
+        }
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+void augment_matching(const matrix::Csr<ValueType, IndexType>* mtx,
+                      array<remove_complex<ValueType>>& weights,
+                      array<remove_complex<ValueType>>& dual_u,
+                      array<remove_complex<ValueType>>& distance,
+                      array<IndexType>& permutation,
+                      array<IndexType>& inv_permutation,
+                      array<IndexType>& unmatched_rows,
+                      array<IndexType>& parents, array<IndexType>& generation,
+                      array<IndexType>& marked_cols,
+                      array<IndexType>& matched_idxs,
+                      remove_complex<ValueType> tolerance)
+{
+    const auto host_exec = mtx->get_executor();
```

Comment:
```suggestion
    const auto host_exec = mtx->get_executor()->get_master();
```
or just `exec` and let the addressable PQ handle getting the host executor.

Reply:
the matrix is already on the host, I changed the name to make it clearer.